lib/auth/integration/integrationv1/service.go

```diff
@@ -403,3 +426,51 @@ func getIntegrationMetadata(ig types.Integration) (apievents.IntegrationMetadata
 func (s *Service) DeleteAllIntegrations(ctx context.Context, _ *integrationpb.DeleteAllIntegrationsRequest) (*emptypb.Empty, error) {
 	return nil, trace.BadParameter("DeleteAllIntegrations is deprecated")
 }
+
+func (s *Service) ensureNoAssociatedResources(ctx context.Context, ig types.Integration) error {
+	switch ig.GetSubKind() {
+	case types.IntegrationSubKindGitHub:
+		return trace.Wrap(s.ensureNoGitHubAssociatedResources(ctx, ig))
+	default:
+		// TODO support this check for other types.
+		return nil
+	}
+}
+
+func (s *Service) ensureNoGitHubAssociatedResources(ctx context.Context, ig types.Integration) error {
+	s.logger.DebugContext(ctx, "Checking GitHub integration associated resources", "integration", ig.GetName())
+	return trace.Wrap(clientutils.IterateResources(ctx, s.backend.ListGitServers, func(server types.Server) error {
+		if server.GetGitHub() != nil && server.GetGitHub().Integration == ig.GetName() {
+			return trace.BadParameter("git servers associated with integration %s must be deleted first", ig.GetName())
+		}
+		return nil
+	}))
+}
+
+func (s *Service) deleteAssociatedResources(ctx context.Context, authCtx *authz.Context, ig types.Integration) error {
+	switch ig.GetSubKind() {
+	case types.IntegrationSubKindGitHub:
+		return trace.Wrap(s.deleteGitHubAssociatedResources(ctx, authCtx, ig))
+	default:
+		return trace.NotImplemented("DeleteAssociatedResources not supported for integration kind %q", ig.GetKind())
+	}
+}
+
+func (s *Service) deleteGitHubAssociatedResources(ctx context.Context, authCtx *authz.Context, ig types.Integration) error {
+	s.logger.DebugContext(ctx, "Deleting git servers associated with integration", "integration", ig.GetName())
+
+	// This RPC only attempts to delete the git server (it's not returning the
+	// git server for the caller to use), so check for types.VerbDelete and
+	// types.VerbList but skip types.Read and authCtx.Checker.CheckAccess on the
+	// resource.
+	if err := authCtx.CheckAccessToKind(types.KindGitServer, types.VerbDelete, types.VerbList); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(clientutils.IterateResources(ctx, s.backend.ListGitServers, func(server types.Server) error {
+		if server.GetGitHub() == nil || server.GetGitHub().Integration != ig.GetName() {
+			return nil
+		}
+		return trace.Wrap(s.backend.DeleteGitServer(ctx, server.GetName()))
```

Comment:
How many git servers typically exist per-integration?

This is doing an individual backend delete operation for each one. If there are many, should we do a range delete to delete them all in a single operation?

Reply:
just one per integration. not expecting a lot integrations/git servers in total either.

Reply:
Ok, this should be fine then.